template/template.go

```diff
@@ -0,0 +1,205 @@
+package template
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+
+	ast "go.expect.digital/mf2/parse"
+)
+
+// MessageFormat2 Errors as defined in the specification.
+//
+// https://github.com/unicode-org/message-format-wg/blob/122e64c2482b54b6eff4563120915e0f86de8e4d/spec/errors.md
+var (
+	ErrSyntax                = errors.New("syntax error")
+	ErrUnresolvedVariable    = errors.New("unresolved variable")
+	ErrUnknownFunction       = errors.New("unknown function reference")
+	ErrDuplicateOptionName   = errors.New("duplicate option name")
+	ErrUnsupportedExpression = errors.New("unsupported expression")
+	ErrFormatting            = errors.New("formatting error")
+)
+
+type ExecFn func(operand any, opts map[string]any) (string, error)
+
+type Template struct {
+	ast       *ast.AST
+	execFuncs map[string]ExecFn
+}
+
+// AddFunc adds a function to the template's function map.
+func (t *Template) AddFunc(name string, f ExecFn) {
+	t.execFuncs[name] = f
+}
+
+func (t *Template) Parse(input string) (*Template, error) {
```

Comment:
Should we error on a function that is not declared in a Template but present in Message Format 2?

Reply:
> To be conformant with MessageFormat 2.0, an implementation MUST implement all of the formatting and selection functions described in the default registry, including all of the options and option values, operands and outputs described by the default registry.

https://github.com/unicode-org/message-format-wg/blob/122e64c2482b54b6eff4563120915e0f86de8e4d/spec/registry.md#conformance-and-use

Reply:
Out of scope of this PR, but we must add all Functions that are defined in 
https://github.com/unicode-org/message-format-wg/blob/122e64c2482b54b6eff4563120915e0f86de8e4d/spec/registry.xml

Reply:
The answer is no. The functions can be added after parsing which is a valid use case.

```go
if t, err = t.Parse("..."); err != nil {
}

t.AddFunc("person", ...)

```

Reply:
okay now i understand what you meant